pkg/kvstore/store/store.go

```diff
@@ -0,0 +1,395 @@
+// synchronous synchronization was successful
+func (s *SharedStore) UpdateLocalKeySync(key LocalKey) error {
+	s.UpdateLocalKey(key)
+
+	if err := s.syncLocalKey(key); err != nil {
+		s.DeleteLocalKey(key)
+		return err
+	}
+
+	return nil
+}
+
+// deleteLocalKey must be called with s.mutex held
+func (s *SharedStore) deleteLocalKey(key LocalKey) {
+	err := kvstore.Delete(s.keyPath(key))
+
+	name := key.GetKeyName()
+	if _, ok := s.localKeys[name]; ok {
+		delete(s.localKeys, name)
+		key.OnDelete()
+
+		if err != nil {
+			s.getLogger().WithError(err).Warning("Unable to delete key in kvstore")
+		}
+	}
+}
+
+// DeleteLocalKey removes a key from bein synchronized with the kvstore
+func (s *SharedStore) DeleteLocalKey(key LocalKey) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.deleteLocalKey(key)
+}
+
+// getLocalKeys returns all local keys
+func (s *SharedStore) getLocalKeys() []Key {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	keys := make([]Key, len(s.localKeys))
+	idx := 0
+	for _, key := range s.localKeys {
+		keys[idx] = key
+		idx++
+	}
+
+	return keys
+}
+
+// getSharedKeys returns all shared keys
+func (s *SharedStore) getSharedKeys() []Key {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	keys := make([]Key, len(s.sharedKeys))
+	idx := 0
+	for _, key := range s.sharedKeys {
+		keys[idx] = key
+		idx++
+	}
+
+	return keys
+}
+
+func (s *SharedStore) getLogger() *logrus.Entry {
+	return log.WithFields(logrus.Fields{
+		"storeName": s.name,
+	})
+}
+
+func (s *SharedStore) updateKey(name string, value []byte) error {
+	s.mutex.Lock()
+
+	// shared key is seen for the first time
+	if s.sharedKeys[name] == nil {
+		if s.localKeys[name] != nil {
+			// if local key, reuse key instance
+			s.sharedKeys[name] = s.localKeys[name]
+		} else {
+			// allocate key for keys from collaborators
+			s.sharedKeys[name] = s.conf.KeyCreator()
+		}
+	}
+
+	err := json.Unmarshal(value, s.sharedKeys[name])
+	s.mutex.Unlock()
+
+	if err != nil {
+		return err
+	}
+
+	s.sharedKeys[name].OnUpdate()
+	return nil
+}
+
+func (s *SharedStore) deleteKey(name string) {
+	s.mutex.Lock()
+	existingKey := s.sharedKeys[name]
+	delete(s.sharedKeys, name)
+	s.mutex.Unlock()
+
+	if existingKey == nil {
+		s.getLogger().WithField("key", name).
+			Warning("Unable to find deleted key in local state")
+		return
+	}
+
+	existingKey.OnDelete()
+}
+
+func (s *SharedStore) listAndStartWatcher() error {
+	listDone := make(chan bool)
+
+	go s.watcher(listDone)
+
+	select {
+	case <-listDone:
+	case <-time.After(listTimeoutDefault):
+		return fmt.Errorf("Time out while retrieving initial list of objects from kvstore")
+	}
+
+	return nil
+}
+
+func (s *SharedStore) watcher(listDone chan bool) {
+	watcher := kvstore.ListAndWatch(s.name+"-watcher", s.conf.Prefix, watcherChanSize)
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+
+			if event.Typ == kvstore.EventTypeListDone {
```

Comment:
Is it guaranteed that you don't receive 2 `kvstore.EventTypeListDone`? Otherwise you will close the `listDone` channel twice, which will cause a panic on the 2nd close.